appengine/standard/firebase/firetactoe/firetactoe.py

```diff
@@ -0,0 +1,270 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Tic Tac Toe with the Firebase API"""
+
+import datetime
+from functools import wraps
+import json
+import os
+import re
+import urllib
+
+from Crypto.PublicKey import RSA
+import flask
+from flask import request
+from google.appengine.api import users
+from google.appengine.ext import db
+import jwt
+import requests
+import requests_toolbelt.adapters.appengine
+
+# Use the App Engine Requests adapter. This makes sure that Requests uses
+# URLFetch.
+requests_toolbelt.adapters.appengine.monkeypatch()
+
+
+# Firebase database server authorization key
+# This is used to authenticate the server and authorize write access
+#
+# Replace this value with your Firebase db key from:
+#
+#     https://console.firebase.google.com/project/_/settings/database
+#
+FIRE_DB_KEY = 'REPLACE_WITH_YOUR_DB_KEY'
+
+_FIREBASE_CONFIG_TEMPLATE = '_firebase_config.html'
+_SERVICE_ACCOUNT_FILENAME = 'credentials.json'
+
+_CWD = os.path.dirname(__file__)
+_IDENTITY_ENDPOINT = ('https://identitytoolkit.googleapis.com/'
+                      'google.identity.identitytoolkit.v1.IdentityToolkit')
+# Firebase database url/channels memoization
+_FIRE_URL = None
+
+
+app = flask.Flask(__name__)
+
+
+def _get_firebase_config():
+    """Parses the client-side Firebase config into a dict."""
+    with open(os.path.join(_CWD, 'templates', _FIREBASE_CONFIG_TEMPLATE)) as f:
+        html = f.read()
+    obj = re.search(r'{([^}]+)}', html, re.MULTILINE)
+    params = [p.strip().split(':', 1) for p in obj.group(1).split(',')]
+    return dict((k.strip('" '), v.strip('" ')) for k, v in params)
+
+
+def _send_firebase_message(u_id, message=None):
+    global _FIRE_URL
+    if not _FIRE_URL:
+        config = _get_firebase_config()
+        _FIRE_URL = '{}/channels/{{}}.json?auth={}'.format(
+            config['databaseURL'], FIRE_DB_KEY)
+
+    if message:
+        return requests.patch(
+            _FIRE_URL.format(u_id), data=message).json()
+    else:
+        return requests.delete(_FIRE_URL.format(u_id)).json()
+
+
+def create_custom_token(uid):
+    """Create a secure token for the given id.
+
+    This method is used to create secure custom tokens to be passed to clients
+    it takes a unique id (uid) that will be used by Firebase's security rules
+    to prevent unauthorized access. In this case, the uid will be the channel
+    id which is a combination of user_id and game_key
+    """
+    with open(os.path.join(_CWD, _SERVICE_ACCOUNT_FILENAME), 'r') as f:
+        credentials = json.load(f)
+
+    payload = {
+        'iss': credentials['client_email'],
+        'sub': credentials['client_email'],
+        'aud': _IDENTITY_ENDPOINT,
+        'uid': uid,
+    }
+    exp = datetime.timedelta(minutes=60)
+    return jwt.generate_jwt(
+        payload, RSA.importKey(credentials['private_key']), 'RS256', exp)
+
+
+class Wins():
```

Comment:
Just

```
X_WIN_PATTERNS = ...
```

It's fine to have these as constants because, well, they are constants. There's no argument for having them in a class. Classes are for collective state or behavior. You could maybe argue for having them in a separate module.